modules/ml/src/main/java/org/apache/ignite/ml/trainers/group/BaseLocalProcessorJob.java

```diff
@@ -0,0 +1,154 @@
+import org.apache.ignite.IgniteCache;
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.Ignition;
+import org.apache.ignite.cache.affinity.Affinity;
+import org.apache.ignite.compute.ComputeJob;
+import org.apache.ignite.ml.math.functions.IgniteBinaryOperator;
+import org.apache.ignite.ml.math.functions.IgniteFunction;
+import org.apache.ignite.ml.math.functions.IgniteSupplier;
+
+/**
+ * Base job for group training.
+ * It's purpose is to apply worker to each element (cache key or cache entry) of given cache specified
+ * by keySupplier. Worker produces {@link ResultAndUpdates} object which contains 'side effects' which are updates
+ * needed to apply to caches and computation result.
+ * After we get all {@link ResultAndUpdates} we merge all 'update' parts of them for each node
+ * and apply them on corresponding node, also we reduce all 'result' by some given reducer.
+ *
+ * @param <K> Type of keys of cache used for group trainer.
+ * @param <V> Type of values of cache used for group trainer.
+ * @param <T> Type of elements to which workers are applier.
+ * @param <R> Type of result of worker.
+ */
+public abstract class BaseLocalProcessorJob<K, V, T, R extends Serializable> implements ComputeJob {
+    /**
+     * UUID of group training.
+     */
+    protected UUID trainingUUID;
+
+    /**
+     * Worker.
+     */
+    protected IgniteFunction<T, ResultAndUpdates<R>> worker;
+
+    /**
+     * Supplier of keys determining elements to which worker should be applied.
+     */
+    protected IgniteSupplier<Stream<GroupTrainerCacheKey<K>>> keySupplier;
+
+    /**
+     * Operator used to reduce results from worker.
+     */
+    protected IgniteBinaryOperator<R> reducer;
+
+    /**
+     * Identity for reducer.
+     */
+    protected final R identity;
+
+    /**
+     * Name of cache used for training.
+     */
+    protected String cacheName;
+
+    /**
+     * Construct instance of this class with given arguments.
+     *
+     * @param worker Worker.
+     * @param keySupplier Supplier of keys.
+     * @param reducer Reducer.
+     * @param identity Identity for reducer.
+     * @param trainingUUID UUID of training.
+     * @param cacheName Name of cache used for training.
+     */
+    public BaseLocalProcessorJob(
+        IgniteFunction<T, ResultAndUpdates<R>> worker,
+        IgniteSupplier<Stream<GroupTrainerCacheKey<K>>> keySupplier,
+        IgniteBinaryOperator<R> reducer,
+        R identity,
+        UUID trainingUUID, String cacheName) {
+        this.worker = worker;
+        this.keySupplier = keySupplier;
+        this.identity = identity;
+        this.reducer = reducer;
+        this.trainingUUID = trainingUUID;
+        this.cacheName = cacheName;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void cancel() {
+
+    }
+
+    /** {@inheritDoc} */
+    @Override public R execute() throws IgniteException {
+        List<ResultAndUpdates<R>> resultsAndUpdates = toProcess().
```

Comment:
Just a question about style. Do we put dot at the end of a line, not start of the next?